### PR TITLE
[CXX-2625] Bring our own make_unique

### PR DESCRIPTION
**Pull request by vector-of-bool**
Refer: CXX-2625. This changeset replaces our use of boost's `make_unique` with our own implementation. This also includes some aside changes related to CMake target scripting.

### CODE REVIEW

File: src/bsoncxx/include/bsoncxx/v_noabi/bsoncxx/stdx/make_unique.hpp [hunk 1]
```diff
@@ -14,52 +14,124 @@
...
+ * @brief Create a new std::unique_ptr that points-to a default-initialized instance of the given
+ * type.
+ *
+ * @tparam T A non-array object type or an array of unknown bound
+ * @param args If T is an object type, then args... must no arguments are allowed.
+ * If T is an array of unknown bound, then args... must be a single size_t specifying
+ * the length of the array to allocate.
+ *
+ * Requires:
+ * - T must be value-initializable
```
Comment (kevinAlbs) [inline src/bsoncxx/include/bsoncxx/v_noabi/bsoncxx/stdx/make_unique.hpp:120]:
```suggestion
 * - T must be default-initializable
```

**Review by kevinAlbs — APPROVED** [review]
Impressive. LGTM.

File: src/bsoncxx/test/make_unique.test.cpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+namespace {
+
+struct something {
+    something(int val) : value(val) {}
+    int value;
+};
+
+TEST_CASE("Create a unique_ptr") {
+    auto ptr = bsoncxx::stdx::make_unique<int>(12);
+    CHECKED_IF(ptr) {
```
[the remark this answers is not in the document]
Comment (vector-of-bool) [reply src/bsoncxx/test/make_unique.test.cpp:17]: I was going to say that this is desired behavior, but I double-checked [the docs](https://github.com/catchorg/Catch2/blob/devel/docs/other-macros.md#assertion-related-macros) and

> `CHECKED_X` macros were changed to not count as failure in Catch2 3.0.1.

:neutral_face:

As someone that has used these macros heavily across dozens projects for several years and seems to have missed this release note, this is an *extremely* breaking change from Catch2, and there doesn't appear to be a viable alternative?? It seems that I have _a lot_ of tests to go and fix now. I need to go lay down.

**Review by kkloberdanz — APPROVED** [review]
Awesome!

**Pull request merged**